pkg/evpn/svi.go

```diff
@@ -27,88 +27,122 @@ import (
-	// TODO: take Port from proto instead of hard-coded
-	vxlan := &netlink.Vxlan{LinkAttrs: netlink.LinkAttrs{Name: resourceID}, VxlanId: int(vxlanid), Port: 4789, Learning: false, SrcAddr: myip}
-	if err := netlink.LinkAdd(vxlan); err != nil {
-		fmt.Printf("Failed to create link: %v", err)
-		return nil, err
-	}
-	// TODO: gap... instead of VPC we are looking for Subnet/Bridge here...
-	if in.Tunnel.Spec.VpcNameRef != "" {
-		// Validate that a Subnet/Bridge resource name conforms to the restrictions outlined in AIP-122.
-		if err := resourcename.Validate(in.Tunnel.Spec.VpcNameRef); err != nil {
-			log.Printf("error: %v", err)
-			return nil, err
-		}
-		// now get Subnet/Bridge to plug this vxlan into
-		bridge, ok := s.Bridges[in.Tunnel.Spec.VpcNameRef]
-		if !ok {
-			err := status.Errorf(codes.NotFound, "unable to find key %s", in.Tunnel.Spec.VpcNameRef)
-			log.Printf("error: %v", err)
-			return nil, err
-		}
-		brdev, err := netlink.LinkByName(path.Base(bridge.Name))
-		if err != nil {
-			err := status.Errorf(codes.NotFound, "unable to find key %s", bridge.Name)
-			log.Printf("error: %v", err)
+	bridge, err := netlink.LinkByName(tenantbridgeName)
+	if err != nil {
+		err := status.Errorf(codes.NotFound, "unable to find key %s", tenantbridgeName)
+		log.Printf("error: %v", err)
+		return nil, err
+	}
+	// Validate that a LogicalBridge resource name conforms to the restrictions outlined in AIP-122.
+	if err := resourcename.Validate(in.Svi.Spec.LogicalBridge); err != nil {
+		log.Printf("error: %v", err)
+		return nil, err
+	}
+	// now get LogicalBridge object to fetch VID field
+	bridgeObject, ok := s.Bridges[in.Svi.Spec.LogicalBridge]
+	if !ok {
+		err := status.Errorf(codes.NotFound, "unable to find key %s", in.Svi.Spec.LogicalBridge)
+		log.Printf("error: %v", err)
+		return nil, err
+	}
+	vid := uint16(bridgeObject.Spec.VlanId)
+	// Example: bridge vlan add dev br-tenant vid <vlan-id> self
+	if err := netlink.BridgeVlanAdd(bridge, vid, false, false, true, false); err != nil {
+		fmt.Printf("Failed to add vlan to bridge: %v", err)
+		return nil, err
+	}
+	// Example: ip link add link br-tenant name <link_svi> type vlan id <vlan-id>
+	vlandev := &netlink.Vlan{
+		LinkAttrs: netlink.LinkAttrs{
+			Name:        resourceID,
+			ParentIndex: bridge.Attrs().Index,
+		},
+		VlanId: int(bridgeObject.Spec.VlanId),
+	}
+	// Example: ip link set <link_svi> addr aa:bb:cc:00:00:41
+	if len(in.Svi.Spec.MacAddress) > 0 {
+		if err := netlink.LinkSetHardwareAddr(vlandev, in.Svi.Spec.MacAddress); err != nil {
+			fmt.Printf("Failed to set MAC on link: %v", err)
 			return nil, err
 		}
-		if err := netlink.LinkSetMaster(vxlan, brdev); err != nil {
-			fmt.Printf("Failed to add vxlan to bridge: %v", err)
+	}
+	// Example: ip address add <svi-ip-with prefixlength> dev <link_svi>
+	for _, gwip := range in.Svi.Spec.GwIpPrefix {
+		fmt.Printf("Assign the GW IP address %v to the SVI interface %v", gwip, vlandev)
+		myip := make(net.IP, 4)
```

Comment:
do we need to create an issue to make the code generic enough to handle ipv4 & ipv6 addresses?

Reply:
yes

Reply:
done